### ISSUE TITLE
Possible issue when nesting thunks

### ISSUE THREAD

**Issue by 4xrsJCr9** [paldepind/snabbdom#40]
Hello,

I am using a thunk-like vnode to emulate components with their own local state. It is common for a thunk-like component to have child components (at an arbitrary depth) which are also the same type of  thunk-like components.

Adding and updating vnodes work correctly, but the ```removeVnodes``` function does not handle the nested thunk-like components correctly. It seems that the ```removeVnodes``` and ```invokeDestroyHook``` functions would have to be updated to correctly handle nested thunks.

Having a thunk as the root vnode also causes trouble in the ```patch``` function.

I also suspect that the vnode returned by the function passed to the ```thunks``` function does not have its ```prepatch/postpatch``` hooks called.

Unfortunately, I am using clojurescript, so I cannot send you a minimal test case (or be a 100% certain that the issue exists at all when using javascript).

**Comment by paldepind** [comment]
Thank you for reporting this issue.

> Unfortunately, I am using clojurescript, so I cannot send you a minimal test case (or be a 100% certain that the issue exists at all when using javascript).

Can you show what your virtual node tree looks like? Are you using the `h` function in your ClojureScript?

**Comment by 4xrsJCr9** [comment]
I was not using the ```h``` function; just some macros that would generate the object literals directly.

I have had to abandon this approach for now (approaching deadline). But, I think you can easily see the problem if you create a thunk to wrap a function that also returns a thunk. Since the ```createElm``` function does not call the ```init``` hook on ```data.vnode``` of the parent thunk vnode, the nested vnode is not created correctly. 

The same thing happens in the ```patchVnode``` function as well.

Modifying the ```createElm``` and ```patchVnode``` to recursively resolve to non-thunk vnodes looks quite easy. However, all other functions would have to start checking for the absence of a DOM node in the ```elm``` property of every vnode before they apply any changes to the DOM. If the ```elm``` property is ```undefined```, you would still need to invoke the relevant lifecycle hooks and then recursively invoke the same process on ```data.vnode```.

Having said that, perhaps you could consider having special vnodes for thunks (or thunk-like objects) which only support the ```init```, ```prepatch```, and ```destroy``` hooks and are never passed to any of the modules.

**Comment by sprat** [comment]
I'm also encountering issues with thunks. I've written a bunch of failing tests showing the bugs I've encountered: see https://github.com/paldepind/snabbdom/pull/56. I can probably fix them too if you want.

**Comment by sprat** [comment]
I am working on the fixes, stay tuned ;)

**Comment by sprat** [comment]
Fixes done, see https://github.com/paldepind/snabbdom/pull/56. Please note that nested thunks are still not addressed in my pull request

**Comment by AlexGalays** [comment]
Any news? Nested thunks seem mandatory.

**Comment by AlexGalays** [comment]
Wrote a failing test:   https://gist.github.com/AlexGalays/cb8742a8f5cf0d179b24

The error is  "childElm is undefined inside createRmCb"; because createRmCb was invoked with an undefined `ch.elm`

**Comment by paldepind** [comment]
@AlexGalays Thank you for the test! I hope to take a proper look at this soon.

**Comment by AlexGalays** [comment]
If you're busy with something else, I could take a look. I'm considering using snabbdom to build a  lightweight and faster alternative to react+flux. I did it in the past with virtual-dom but nowadays its performances are too close to React's for me to care. On the other hand, I don't yet have a good understanding of snabbdom's stability. Do you use it in production?

**Comment by paldepind** [comment]
I _might_ have an idea for a different thunks implementation that should be simpler and support nesting. I'll try implementing it and see how it goes.

> On the other hand, I don't yet have a good understanding of Snabbdom's stability. Do you use it in production?

I have nothing in production but I've used it on a few personal projects. Other people use it in production though. There has been a bunch of bugs/inconveniences discovered by people with different use cases than mine. So it's somewhat ironed out. I'd say it's pretty stable :smile: (except thunks).

**Comment by AlexGalays** [comment]
With version 0.3.4, my test now passes btw!
It appears directly nested thunks do not work, but thunks separated by at least one layer do.

For instance if you slightly modify sprat's test, it can be made to work : https://gist.github.com/AlexGalays/10baf1a96464e1bd8346

Of course, I think it should always work; a thunk directly under a thunk is a scenario I've encountered before.

**Comment by sprat** [comment]
I think the test in your gist is already covered by the `renders correctly child thunk` test. The only test commented is the directly nested thunk test, which is the harder one to address ;)

**Comment by AlexGalays** [comment]
If I modify the beginning of the code in `thunk.js` like so:

```
function init(thunk) {
  thunk.data.vnode = getvnode(thunk);
}

function getvnode(thunk) {
  var cur = thunk.data;
  var vnode = cur.fn.apply(undefined, cur.args);

  return vnode.sel.indexOf('thunk') === 0
    ? getvnode(vnode)
    : vnode;
}
```

The first part of the nested test (the creation) passes; However, the thunk/thunk patching fails at the line `parentElm.insertBefore(elm, oldVnode.elm);` because parentElm is null; I'm not yet sure why. The whole test runs without ever actually attaching elements to the document right?

**Comment by paldepind** [comment]
The thunk issues should be fixed by the latest release 0.5.0.
